File: Network/src/Dispositive.java
import java.util.ArrayList;

public abstract class Dispositive implements Runnable
{
	//Dispositivi che conosce
	protected ArrayList<Dispositive> disp=new ArrayList<Dispositive>();
	//Operazioni che deve fare
	protected ArrayList<Operation> ope=new ArrayList<Operation>();
	//Nome
	protected int name;
	//Numero di operazioni
	protected int operations=0;
	
	public Dispositive(int n)
	{
		name=n;
	}
	
	public void setOperations(int op)
	{
		operations=op;
	}

	public void register(Dispositive dispositive, int j, int i) 
	{
		disp.add(dispositive);
		System.out.println("Dispositivo "+i+" aggiunto a "+j);
	}
	
	public void addOp(Operation o)
	{
		ope.add(o);
	}
	
	//Crea un nuovo thread perche' puo' inviare piu' messaggi assieme
	public void run()
	{
		for (int i=0; i<ope.size(); i++)
		{
			final int num=i;
			System.out.println("[D]Preparazione operazione da "+ope.get(i).getFrom2()+" a "+ope.get(i).getTo2()+"!");
			new Thread() 
			{
				public void run() 
				{
					send(num);
				}
			}.start();
		}
		
		//Quando ha finito, aspetta un po' prima di fermarsi
		try 
		{
			Thread.sleep(5000*operations);
		} 
		catch (InterruptedException e) 
		{
			//e.printStackTrace();
		}
	}
	
	private void send(int i)
	{
		Dispositive to=ope.get(i).getTo();
		String msg=ope.get(i).getMessage();
		try 
		{
			to.receive(msg, ope.get(i).getFrom2(), ope.get(i).getTo2());
		} 
		catch (InterruptedException e) 
		{
			//e.printStackTrace();
		}
	}
	
	//Puo' ricevere piu' messaggi assieme
	abstract public void receive(String msg, int i, int j) throws InterruptedException;
}

File: NetworkS/src/Dispositive.scala
import scala.actors._
import scala.actors.Actor._
import scala.collection.{immutable, mutable}

trait Dispositive extends Actor
{  
	//Operazioni che deve fare
	protected var ope=new mutable.ArrayBuffer[Operation]();
	//Dispositivi che conosce
	protected var disp=new mutable.ArrayBuffer[Dispositive]();
	//Operazioni fatte
	protected var op:Int=0;
	//Nome
	protected var gid: Int=0;

	def id=gid;
	def id_= (i:Int) = {gid=i;};
	
	def register(dispositive: Dispositive, j: Int, i: Int) 
	{
		disp+=dispositive;
		println("Dispositivo "+i+" aggiunto a "+j);
	}
	
	def addOp(o: Operation)
	{
		ope+=o;
	}
	
	//Posso inviare e ricevere assieme, quindi creo un actor per ogni messaggio
	def act() 
	{
		loop
		{
			react
			{
				case o:Operation => actor { receive(o.from2, o.to2); } 
				case k:Int => actor { send(k); }
				case "send" => actor { prepare; }
				case 'Interrupt => exit();
			}
		}
	}
	
	def receive (i: Int, j:Int); 
	
	//Invia un messaggio alla volta con un po' di attesa
	def send(i: Int) =
	{
		val o:Operation=ope(i);
		val to: Dispositive=o.to;
		op+=1;
		Thread.sleep(2000);
		to!o;
	}
	
	//Prepara i messaggi all'invio
	def prepare() =
	{
		for (i<-0 to ope.size-1)
		{
			println("[D] Preparazione messaggio");
			val s:Dispositive=this;
			s!i; 
		}
	}
}